### ISSUE TITLE
PMTUD creates too much overhead in some applications

### ISSUE THREAD

**Issue by huitema** [private-octopus/picoquic#1230]
Path MTU Discovery operates by sending "MTU probe" packets that contain just PING and PAD frames. If a probe of length L gets acknowledged, the code learns that setting the PMTU to L is probably fine. There is logic in Picoquic to try different lengths. The connection only works if the minimum packet size is supported (1232 for IPv6, 1252 for IPv4). If that does not work, the path will not be used at all. Otherwise, the initial PMTU is set to that value. After that, Picoquic will try to send a succession of probes until one works: first try the maximum packet size authorized by the peer, then if that does not work try something intermediate, etc., until either a good size has been found, or it does not make sense to further try.

The problem with that is overhead. The PING+PAD MTU probes are pure overhead. On one hand that's good, because sending those probes does not affect the logic of the application. On the other hand, the overhead creates additional traffic on the links, more packets to process by the peers, etc. Some applications just send a few short messages -- DNS over QUIC would be an example, text messaging would be another. The MTU probes can end up being a big fraction of the total traffic, for little benefit.

The logic in the code checks whether to send PMTUD or not:

* If the transport knows that it will send lots of data, PMTUD is deemed a good investment and the value is set to "required"
* Else it is set to "optional"

If required, the probes are sent as soon as the bandwidth is available. If "optional", they are only sent if there is nothing else to send. But that logic does not help the low-traffic applications, because the probes will still be sent as soon as some network bandwidth becomes available, creating overhead. The reverse problem happens for some high-bandwidth applications, because it might take some time for the transport to decide that the "the application will send lots of data". 

It would probably be nice to have a way to set PMTUD policy, with a few defined behaviors:

* Do PMTUD always, asap -- whether the logic says optional or required.
* Do PMTUD according to the current logic, i.e. send optional packets if the is space.
* Only do PMTUD is there is lots of data to send.
* Never do PMTUD, stick with 1232.

We have one API already defined:
```
void picoquic_cnx_set_pmtud_required(picoquic_cnx_t* cnx, int is_pmtud_required);
```
This API takes a binary parameter. If set to 1, the code will do PMTUD always, which is great for high bandwidth applications. If set to 0, the code will use the current logic. It seems that we need to use two other values for "only if there is lots of data to send" and "never".

The API operates at the individual connection level. That's useful, as it allows changing the behavior of ongoing connections, but in many applications it might be simpler to define the behavior at the QUIC context level.

**Comment by huitema** [comment]
Another issue with PMTUD is that it might be too optimistic, trying large sizes like 1536 bytes or even 8K bytes in the off chance that it might work. The algorithm will eventually converge to a reasonable MTU, but starting with a large size will create extra overhead. In many deployments, it might make sense to limit PMTUD to only try the MTU values most likely to be supported on the most common paths. For example, if the application is optimized for running on cellular links, it might make sense to limit the PMTUD discovery to an UDP payload size of 1384 bytes (for IPv4) or 1364 bytes (for IPv6). That's the purpose of the API:
```
void picoquic_set_mtu_max(picoquic_quic_t* quic, uint32_t mtu_max);
```
There may be some confusion between usage of that API and usage of the transport parameter `max_packet_size`, which may also be set by the application. These are however not equivalent. Setting the `max_packet_size` in the local parameters only limits what the remote peer can send to the local peer, but not in the other direction. Suppose the desired result is "even if the remote peer accepts packets of 8K bytes, only negotiate up to 1364 bytes". The only way to do that is through the API `picoquic_set_mtu_max`. The behavior should probably be:

1) If the application sets the `max_packet_size` transport parameter explicitly, send that value to the peer.
2) If the application does not set the parameter, set the value to `mtu_max` and send that to the peer. 
3) If the peer sends a  `max_packet_size` value larger than `mtu_max`, set the PMTUD limit to `mtu_max`.

**Comment by davidfdzp** [comment]
Very interesting considerations. How are these packet sizes you mention determined?
"The connection only works if the minimum packet size is supported (1232 for IPv6, 1252 for IPv4)."
"if the application is optimized for running on cellular links, it might make sense to limit the PMTUD discovery to an UDP payload size of 1384 bytes (for IPv4) or 1364 bytes (for IPv6)."
In a Windows computer, using Ethernet interface, I understand it would be 1472 bytes for IPv4, because of this:
C:\>ping techrepublic.com -f -l 1472

Pinging techrepublic.com [107.178.248.185] with 1472 bytes of data:
Reply from 107.178.248.185: bytes=68 (sent 1472) time=22ms TTL=116

But then:
C:\>ping techrepublic.com -f -l 1473

Pinging techrepublic.com [107.178.248.185] with 1473 bytes of data:
Packet needs to be fragmented but DF set.

These days of remote working, when connected to office via VPN the MTU is 1332 bytes:
C:\>ping techrepublic.com -f -l 1332

Pinging techrepublic.com [107.178.248.185] with 1332 bytes of data:
Request timed out.

C:\>ping techrepublic.com -f -l 1333

Pinging techrepublic.com [107.178.248.185] with 1333 bytes of data:
Packet needs to be fragmented but DF set.

I understand that depending on the VPN used the MTU will end up having different values, not so easy then.

Tested in bash with, e.g.: ping -s 1500 -M do example.com

**Comment by victorstewart** [comment]
@davidfdzp 
those values are related some cellular values i was discussing: the MTU for LTE is 1428, but some carriers run at 1420. then 5G adds another 8 bytes. so on mobile you’re looking at either 1412, 1420 or 1428. 

as you mentioned, problem is virtualization layers in between though. maybe the client is running a VPN, which can have widely varying header sizes depending on protocol and encryption algo + # of bits. or the server might be running in a VM or behind some container defined network or VPN.

so another issue in general is the optimal increments and search space.

**Comment by huitema** [comment]
Issues were fixed in PR #1231
